bindgen-tests/tests/headers/issue-2646.h

```diff
@@ -0,0 +1,30 @@
+// bindgen-flags: --rustified-enum 'Plain.*' --rustified-enum 'TryFromRaw.*=try_from_raw' --rustified-enum='FromRawUnchecked.*=from_raw_unchecked' --rustified-enum='Both.*=try_from_raw,from_raw_unchecked' --rustified-enum 'NonExhaustive.*=non_exhaustive'
+
+enum Plain {
+    Plain1,
+    Plain2,
+    Plain3
+};
+
+enum TryFromRaw {
+    TFR1 = -1,
+    TFR2 = 5,
+    TFR3
+};
+
+enum FromRawUnchecked {
+    FRU1 = 6,
+    FRU2 = 10,
+    FRU3 = 11,
+};
+
+enum Both {
+    Both1,
+    Both2 = -1,
+    Both3,
+};
+
```

Comment:
It could be a good idea to add a few functions here, e.g. `enum Plain f(enum Plain);` and so on, to double-check the generated type in `expectations`.

Reply:
Oof, it appears after adding some tests that it is converting it to the type that will result in undefined behavior. I'll have to take a look at how to avoid that. I think the code to handle that is in a place I haven't touched, so I'll have to track down where that's happening.

Reply:
That is what I feared, yeah. Thanks!

Reply:
So is what you're recommending only doing this in the case of TryFrom generation like with the constants?

Reply:
I am not sure what you mean -- do you have an example?

Reply:
I think what I'm getting at here is that it is bindgen's current behavior is to use the enum type as the function argument. I think it is unlikely to get merged if I change the behavior in the default case to require an integer when it used to be an enum. That to me sounds like a breaking change and would probably make a lot of people angry. I'm basically suggesting that (if possible) we only make the argument an integer if `TryFrom` is also implemented on the enum type as that feature has not been released so it should theoretically be backwards compatible. I'll need to dig into this more but I believe that could be possible.

Reply:
I think that receiving the integer instead of the enum type would be a reasonable change if it's gated behind the `TryFrom` implementation for the reasons @jbaublitz mentions: It is backwards compatible.

However, from a personal perspective, translating `f(enum Plain)`  as `f(plain: int)` instead of `f(plain: Plain)` it is the most "correct" translation as the C side can actually pass values that aren't defined inside `Plain` anyway. It would force the Rust side to handle this case explicitly in one way or another.

Reply:
> I'm basically suggesting that (if possible) we only make the argument an integer if `TryFrom` is also implemented on the enum type as that feature has not been released so it should theoretically be backwards compatible.

Yeah, definitely, the idea was to have a new mode/flag/whatever, i.e. not to break existing users. Having said that, I am not sure changing the behavior of the the type used for function signatures and so on based on `TryFrom` is intuitive. Perhaps naively I always thought of this as a new mode altogether, or perhaps as a variation of constified, rather than of rustified.

@pvdrz I agree that it would be better to generate things differently to avoid UB in some of the existing modes, but does `bindgen` generally break existing users? What is the usual policy? Do you mean deprecating existing modes could be an option? Are a lot of projects out there using rustified enums?

Reply:
I'm personally open to changing it to an int in whatever cases make sense to @pvdrz.

Reply:
> does `bindgen` generally break existing users? What is the usual policy? 

That's a muddy topic. Strictly speaking, we're a pre-1.0 crate so that means that breaking changes are expected. At the same time, we try to avoid them as much as possible unless there's a strong reason to do so

> Do you mean deprecating existing modes could be an option? 

Yes I'd say this is something on the table. Specially if we can first deprecate and then remove after a while.

> Are a lot of projects out there using rustified enums?

Unknown, we don't have a lot of metrics on this, we could try to check which crates on crates.io depend on bindgen but that leaves out all the projects that just use the CLI or projects that aren't on crates.io

Reply:
Maybe, one alternative to roll this change slowly would be to add another flag like `--use-enum-ctypes-on-fns` (name is a WIP :rofl:) that when enabled, makes all functions use the ctype instead of the rustified enum. We could keep it disabled by default and eventually enable it.

@jbaublitz @ojeda  what do you think?

Reply:
How would a user call `takes_plain` with a value that is not in the Rust-side `enum`? i.e. a C function may need to be use with more values too. Of course, perhaps the `bindgen` user should avoid this kind of `enum` in this case.

Reply:
> How would a user call `takes_plain` with a value that is not in the Rust-side `enum`? i.e. a C function may need to be use with more values too. Of course, perhaps the `bindgen` user should avoid this kind of `enum` in this case.

I think the only case where a Rust user would call `takes_plain`  with a value outside the rustified enum would be if there is a common pattern in the C library like:
```c
// This value could be outside of `Plain`
enum Plain value = returns_plain();
// But that's OK because this function already assumes that this could happen and preemptively checks that inside
takes_plain(value);
```

So doing that from the rust side would be weird as you'd have:

```rust
unsafe extern "C" {
   fn returns_plain() -> Plain_ctype;
   fn takes_plain(plain: Plain);
}
// This value could be outside of `Plain`
let value: Plain_ctype = unsafe { returns_plain() };
// But here, we have to do an unsafe transmute that might cause UB even though the call to the C function would be fine with `value`
unsafe { takes_plain(std::mem::transmute(value)) };
```
However, if you just use the `Plain_ctype` everywhere, that's not cumbersome anymore

```rust
unsafe extern "C" {
   fn returns_plain() -> Plain_ctype;
   fn takes_plain(plain: Plain_ctype);
}
// This value could be outside of `Plain`
let value: Plain_ctype = unsafe { returns_plain() };
// No weird transmute
unsafe { takes_plain(value) };
```

Reply:
(My previous comment was intended to reply to your "two functions", rather than the `--use-enum-ctypes-on-fns` one; our messages crossed :)

> However, if you just use the `Plain_ctype` everywhere, that's not cumbersome anymore

Exactly, that is why I think using the Rust `enum` for parameters may not be great, unless the expectation is that users avoid it in some cases. So they would probably still not use it as their "default" or "catch all" case, i.e. `--... *`.

> However, I could see a case where you'd want to offer both the rustified enum and safe wrappers to those functions.

Hmm... What do you mean by safe wrappers? i.e. the raw functions would still need to be marked unsafe, even if we took Rust `enum`s, so the user would still need to wrap it.

Having said that, I would like to eventually have a `[[safe]]` attribute on the C side that allows us to do things like that -- for the kernel for instance, we are now at the point were starting to add annotations on the C side for several things may be OK for some subsystems at least. Similarly, we could have a `[[enum ...]]` thing that gives `bindgen` this information without having to maintain the lists outside the source code. I don't know if nowadays `libclang` preserves unknown attributes -- if so, we could already do it without having to modify and wait for Clang.

The other extreme is to have an "easy-to-use" mode, where by default all is provided: the raw C type and its constants, plus a Rustified `enum` type, plus a non-exhaustive one, plus the conversion functions (fallible, infallible and unsafe) to convert to both of those... Then users can use whatever they think it is best for each `enum`, without having to know about the `bindgen` kinds, and it is sound. Then when they are sure how they use each `enum`, they can restrict it to one kind via the arguments (or via attributes if we ever get there) so that they avoid `bindgen` generating code that will not be needed. So it would be easy to start using, and allow people to customize later. But it would be more "expensive" by default.

Reply:
> (My previous comment was intended to reply to your "two functions", rather than the `--use-enum-ctypes-on-fns` one; our messages crossed :)
> 
> > However, if you just use the `Plain_ctype` everywhere, that's not cumbersome anymore
> 
> Exactly, that is why I think using the Rust `enum` for parameters may not be great, unless the expectation is that users avoid it in some cases. So they would probably still not use it as their "default" or "catch all" case, i.e. `--... *`.

Yeah I think that the extra convenience of using the rustified enum would only work if you want to expose the enum itself. And if you want to expose safe bindings for functions that use the c enum, then you have to do those yourself accordingly to the invariants of the library itself.

> > However, I could see a case where you'd want to offer both the rustified enum and safe wrappers to those functions.
> 
> Hmm... What do you mean by safe wrappers? i.e. the raw functions would still need to be marked unsafe, even if we took Rust `enum`s, so the user would still need to wrap it.

That's true, even if the function takes the rustified enum then you still have to figure out its safety requirements and expose it accordingly. So you'd still have to do things like

```rust
mod bindings {
    unsafe extern "C" {
        pub fn takes_plain(plain: Plain);
    }
}

/// Safety: It is safe, believe me.
pub fn takes_plain(plain: Plain) {
    unsafe { bindings::takes_plain(plain) }
}
```

So, having to add an extra `plain.to_raw_ctype()` or whatever to obtain the C value seems like little effort. Maybe @emilio or @kulp have a better idea on why bindings of functions that take C enums become functions that take Rustified enums instead of using the integer type directly as they have been around for longer.
 
> Having said that, I would like to eventually have a `[[safe]]` attribute on the C side that allows us to do things like that -- for the kernel for instance, we are now at the point were starting to add annotations on the C side for several things may be OK for some subsystems at least. Similarly, we could have a `[[enum ...]]` thing that gives `bindgen` this information without having to maintain the lists outside the source code. I don't know if nowadays `libclang` preserves unknown attributes -- if so, we could already do it without having to modify and wait for Clang.

That sounds like a sensible addition. In the meantime, it is already possible to embed html-like tags in the comments of the headers and bindgen can read those to introduce custom behavior.

> The other extreme is to have an "easy-to-use" mode, where by default all is provided: the raw C type and its constants, plus a Rustified `enum` type, plus a non-exhaustive one, plus the conversion functions (fallible, infallible and unsafe) to convert to both of those... Then users can use whatever they think it is best for each `enum`, without having to know about the `bindgen` kinds, and it is sound. Then when they are sure how they use each `enum`, they can restrict it to one kind via the arguments (or via attributes if we ever get there) so that they avoid `bindgen` generating code that will not be needed. So it would be easy to start using, and allow people to customize later. But it would be more "expensive" by default.

Yeah ideally you should be able to decide if a C enum should be translated in different ways without interference between each representation. I think the API would need some bike-shedding, specially from the CLI side where there isn't much flexibility. Maybe the idea of having a `bindgen.toml` file where we can store more structured information would be useful here as you could do something like

```toml
[[enum_style]]
regex = "Foo"
rustified = { generate = true, try_from_raw = true, from_raw_unchecked = true }
constified = true
```

Then anything that matches the `Foo` regex would get all of those immediately. Maybe a CLI equivalent could be `--enum-style "Foo=rustified(try_from_raw,from_raw_unchecked),constified"`

It would also be nice to be able to replace the multitude of `--newtype-enum` `--newtype-global-enum`, `--constified-enum`, etc, by a single option that can be extended. This would also give @jbaublitz some flexibility as it wouldn't be necessary to include the raw type constants with this feature as those would be enabled automatically if you pass the extra `constified` style.